website/docs/index.html.markdown

```diff
@@ -31,13 +33,32 @@ resource "dns_a_record_set" "www" {
 }
 ```
 
+Using GSS-TSIG (RFC 3645):
+
+```hcl
+provider "dns" {
+  update {
+    server   = "ns.example.com" # Using the hostname is important in order for an SPN to match
+    gssapi   = true
+    realm    = "EXAMPLE.COM"
+    username = "user"
+    keytab   = "/path/to/keytab"
```

Comment:
From reading the configuration validation logic earlier, I got the sense that these `realm`, `username`, `password`, and `keytab` arguments are applicable only to `gssapi` mode.

If that is true, what do you think about reorganizing this a little so that these settings are inside a `gssapi` block, like this?:

```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi {
      realm    = "EXAMPLE.COM"
      username = "user"
      keytab   = "/path/to/keytab"
    }
  }
}
```

We tend to use this configuration pattern when activating a feature implies requiring an additional set of arguments that would otherwise not be required. Arranging it this way would mean that the presence of the block activates the feature and all of the nested arguments could be marked as `Required: true`, and that may serve to simplify the validation logic by having Terraform and the provider SDK check these are present automatically.

In retrospect perhaps we ought to have originally put the existing key arguments in a `tsig` block too for symmetry, but I don't think that's important enough to warrant a breaking change now.

Reply:
Yes, I will do that. I think originally there was just a boolean toggle to enable GSSAPI as it picked everything else up from the environment due to the libraries used and then when I swapped the libraries to eliminate the CGo hurdle the parameters just organically grew from there. It should make the parameter validation much simpler, I agree it's a bit hairy now checking for the various allowed permutations.

Reply:
@apparentlymart So the only edge case is that it is possible to enable GSSAPI without any further configuration for it to just use the current environment, so you would in theory end up with something like this:
```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi {}
  }
}
```
Is that valid/acceptable? Alternatively, I could remove that ability so that the credentials are always explicitly configured in the provider.

Reply:
This is a good question! Thanks.

Although it's currently not captured anywhere programmatically, we've noted before that in effect there are two different classes of provider configuration arguments: those that describe where things are to be created, and those that describe which human or automaton is running Terraform in terms that make sense to the remote system. While in most providers we _allow_ both to be set either in configuration or in the environment, we tend to recommend setting the "where"/"how" settings in the configuration but leaving the "who" settings for the environment or some other ambient source.

In this case, the `realm` argument feels to be to be roughly analogous to the AWS provider's `region` and `assume_role` arguments: they are more about where and how to create things than who is doing it. `assume_role` is, similar to `realm`, kinda on the edge of that: it is a _constraint_ on "who", but is still more like a "where"/"how".

With all of that theory out of the way, I'd expect that we'd want to show the primary way of using this with the `realm` set in the configuration but the other two settings coming from the environment:

```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi {
      realm = "EXAMPLE.COM"
    }
  }
}
```

I read the above as "send updates to `ns.example.com`, which requires a credential from the `gssapi` realm `EXAMPLE.COM`". As a _user_, I then need to set up my environment to provide a suitable credential. The required realm is presumably chosen by the remote server and not by the user.

While on this subject, I wonder if it would make sense for the username and the keytab to be picked up from the logged in user by default, if possible. Kerberos and GSSAPI are normally used in "single sign-on" environments, where the user probably logged in to their current shell with Kerberos already and so their system username is probably a good default `username` and perhaps we can find a suitable `keytab` systematically somehow?

I'd still leave it so that all of these _can_ be configured in the environment, since it's pragmatic to give that flexibility for unusual situations, but I think we can design the configuration around a recommended "normal" pattern for what is and is not set, and document with that in mind.

How does all that seem to you? I assume you're working in a Kerberized environment already and so you'll have a more up-to-date sense of what we can and cannot determine automatically from the session; we usually try to make Terraform providers read from the "standard" credentials source for the system in question by default, rather than always requiring Terraform-specific configuration, but it's been so long since I was working in a Kerberos environment that I'm sure I'm forgetting details that are relevant here.

Reply:
I think we've crossed wires slightly, however you've possibly answered the next question I had in mind.

The use case I was talking about was using the currently logged in user, (probably shouldn't have said environment in that case!), if you don't have any configuration (apart from enabling GSSAPI support) _and_ none of the GSSAPI-specific provider environment variables are set then it will try and use a ticket for the current user, (this was initially the only way it would work due to the first GSSAPI library I used having no Kerberos smarts). No other configuration is then needed as it will pick up everything automatically, you just need a way to say "do GSSAPI" hence conceptually you could end up with just an empty `gssapi {}` block in the provider as you still need to enable it.

A keytab can be utilised automatically in the above scenario, provided it has been provided to `kinit` previously. Something like `kinit -F -k -t /path/to/keytab username` IIRC. According to https://web.mit.edu/kerberos/krb5-devel/doc/basic/keytab_def.html#default-client-keytab there are some Kerberos-level defaults that can be tried as well to locate a keytab.

My next question I thought of is that currently the provider allows environment variables for literally every setting, and I was wondering if that is desired or not? Also, the code seems to duplicate the environment variable way of retrieving settings, i.e.:
```golang
"server": {
        Type:        schema.TypeString,
        Required:    true,
        DefaultFunc: schema.EnvDefaultFunc("DNS_UPDATE_SERVER", nil),
},
```
and then further down in `configureProvider()`:
```golang
if len(os.Getenv("DNS_UPDATE_SERVER")) > 0 {
        server = os.Getenv("DNS_UPDATE_SERVER")
} else {
        return nil, nil
}
```
Is this correct?

Also, in the context of using the current user, there would be no way of expressing "do GSSAPI" without some sort of flag given there won't be any other GSSAPI provider environment variables set. So maybe the provider could be made to try GSSAPI if there are no explicit TSIG settings and then fall back to unsigned updates, (i.e. no TSIG of any sort)? What do you think?

Reply:
Looking at this again today with fresh eyes, I see that my initial thought of `realm` effectively being required isn't really true, because in real-world Kerberos environments there can often be multiple realms that are all equally fine to use, separated only for administrative purposes, and so it would be reasonable to have a DNS server that accepts updates from multiple different realms and have the end-user running Terraform be the one to decide. And given that, having it pick up an "ambient" realm for the logged in user makes sense, reducing the configuration to just an empty `gssapi {}` block. An empty block is a little unconventional, but not without precedent and I think worthwhile given the other benefits it gives us in clarifying which settings are GSSAPI-specific.

If there's a reasonable way to reduce the required configuration burden down to _just_ the server in the already-logged-in-with-GSSAPI scenario then I think that would be awesome:

```hcl
provider "dns" {
  update {
    server = "ns.example.com"
  }
}
```

This is consistent with the goal of having the configuration describe _where_ while the _who_ comes from ambient information, and I think that's the ideal case. Since there isn't (as far as I know) a "standard" way to set the TSIG settings ambiently we invented a provider-specific environment variable scheme for those, but I'd always prefer a mechanism that re-uses existing conventions where it's available, so having the provider try to use GSSAPI by default if the TSIG settings are absent seems great to me.

Regarding environment variables for every setting: that tends to be the case in most providers mainly because it makes test automation more convenient. In practice I think we'd expect end-users _not_ to use the `DNS_UPDATE_SERVER` environment variable, but we tend to be pragmatic about these things and rely on recommendations and good examples rather than forcing a particular usage pattern.